amp/parse.go

```diff
@@ -32,6 +41,62 @@ type Size struct {
 	Width          int64
 }
 
+// Policy consent types
+const (
+	ConsentDenied = iota
+	ConsentTCF1
+	ConsentTCF2
+	ConsentUSPrivacy
+)
+
+func ReadPolicy(ampParams Params, req *openrtb2.BidRequest, pbsConfigGDPREnabled bool) (privacy.PolicyWriter, error) {
+
+	if len(ampParams.Consent) == 0 {
+		return privacy.NilPolicyWriter{}, nil
+	}
+
+	var rv privacy.PolicyWriter = privacy.NilPolicyWriter{}
+	var warning error
+	var errMsg string
+
+	switch ampParams.ConsentType {
+	case ConsentTCF2:
+		if pbsConfigGDPREnabled {
+			// Even if consent is invalid, we write the consent string to req.user.ext.consent
+			rv = gdpr.ConsentWriter{ampParams.Consent}
+
+			if tcf2.IsConsentV2(ampParams.Consent) {
```

Comment:
Why do we validate the consent string here if we always put in the request? If we omit this, will downstream code put in a similar warning and/or error?

Reply:
Agree Brian. The sole reason I included a GDPR string validation call is to return a warning. Let me know your take on the new logic.

Reply:
> Why do we validate the consent string here if we always put in the request?

In the discussions we had on how to implement this, we decided to leave invalid strings, issue a warning, and let the `consent` string malformedness end up not granting the permissions downstream.

>  If we omit this, will downstream code put in a similar warning and/or error?

It won't. This is the only warning on this regard. In `exchange/utils.go` we resolve the `gdprApplies` value and parse the permissions found in the consent string, but at no point do we exit with an error if the `consent` string is malformed.